lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -1544,6 +1544,10 @@ public void send(ProtocolMessage msg, boolean queueEvents, CompletionListener li
     private void sendImpl(ProtocolMessage message, CompletionListener listener) throws AblyException {
         if(transport == null) {
```

Comment:
It looks like you're checking against a field (reference) here and then later on using that same reference to call `send` on. Without looking at the threading model in more depth, I would worry that it might become `null` between here and that `send` call. I wonder if it might be better to cache the reference to a local `final` variable at the start of this method's implementation and then use that local throughout.

Reply:
I have not written this part 🤷

Reply:
Indeed, but at the point this was written then this method implementation only read the `transport` field once. You've changed that.

Reply:
> then this method implementation only read the transport field once. You've changed that.

Sorry can you clarify this please? This field was already reading this and returning if the transport was null. Do you mean that this field isn't synchronized  and the state could change after callback and return?

Reply:
Yes, so I am worried about the field value changing from being a valid reference to `null` between your null-check and the subsequent use of the field to call `send`.

Reply:
Of course. But I cannot see how this differs from the method returning after checking the `transport` which can change in the point of return which is also a synchronization issue. What I do here is to provide a path (in this case a completion listener) so that they can have opportunity to respond as this method was originally designed, in a thread confined manner